### PR TITLE
syncer(dm): save table checkpoint after a DDL is filtered

### LINKED ISSUE #5272: after filtered some DDL event and manually fix downstream, tracker can't track table structure
### What did you do?

task

```
...
block-allow-list:        # 上游数据库实例匹配的表的 block-allow-list 过滤规则集，如果 DM 版本 <= v2.0.0-beta.2 则使用 black-white-list
  bw-rule-1:             # 黑白名单配置的名称
    do-dbs: ["test"] # 迁移哪些库
...
filters:
  filter-rule-1:
    schema-pattern: "test"
    table-pattern: "test1"
    events: ["all ddl"]
    action: Ignore
```

1. create `test.test1` in upstream
2. start-task --remove-meta
3. `alter table test1 add column c4 int;` in upstream
4. `create table test2 (c int primary key);` in upstream or wait 30s, to flush checkpoints
5. `insert into test1`. Now task will report error because downstream doesn't have column `c4`
6. `alter table test1 add column c4 int;` in downstream
7. resume-task

### What did you expect to see?

task goes on

### What did you see instead?

gen insert sqls failed, sourceTable: `test`.`test1`, targetTable: `test`.`test1`: Column count doesn't match value count: 3 (columns) vs 4 (values)",

### Versions of the cluster

DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

at least v5.4.0

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: ref #5272 

### What is changed and how it works?

as title

### Check List 

Tests 

 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
save table checkpoint after a DDL is filtered
```

### CODE REVIEW

**Comment by lance6716** [comment]
/hold

test is invalid

File: dm/tests/tracker_ignored_ddl/run.sh [hunk 1]
```diff
@@ -27,19 +27,26 @@ function run() {
...
 	# need operate tidb
 	run_sql_tidb "alter table $TEST_NAME.t1 add column ignore_1 int;"
 
 	# resume task
 	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
 		"resume-task test" \
 		"\"result\": true" 2
-	sleep 1
+	sleep 3
 	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
```
Comment (GMHDBJD) [inline dm/tests/tracker_ignored_ddl/run.sh:50]: Seems the run_dm_ctl_with_retry already meets the requirements

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: dm/syncer/checkpoint.go [hunk 2]
```diff
@@ -128,6 +128,7 @@ func (b *binlogPoint) rollback(schemaTracker *schema.Tracker, schema string) (is
 	b.flushedPoint.location.ResetSuffix()
 	b.savedPoint.location = b.flushedPoint.location
 	if b.savedPoint.ti == nil {
+		// TODO: if we forget to save table info for table checkpoint, this is also nil!
```
Comment (D3Hunter) [inline dm/syncer/checkpoint.go:131]: table point which rollbacks before can also be nil

File: dm/syncer/syncer.go [hunk 3]
```diff
@@ -3312,6 +3312,9 @@ func (s *Syncer) loadTableStructureFromDump(ctx context.Context) error {
 					zap.Error(err))
 				setFirstErr(err)
 			}
+			// TODO: we should save table checkpoint here, but considering when
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:3315]: here too
https://github.com/pingcap/tiflow/blob/be428b781aa9cdb67f78db04bc04d9e8670f1b72/dm/syncer/syncer.go#L806

suppose a incremental task:
- upperstream `create table t(c1 int primary key, c2 int)` and insert some rows
- downstream mistake created as `create table t(c1 int primary key)`
- start task
- task fail and table point rollback, and auto resume will use downstream ddl and exec it into schema tracker
- downstream add c2
- will fail now.

Comment (lance6716) [reply dm/syncer/syncer.go:3315]: I'll address comment and add a test about this case tomorrow

File: dm/tests/tracker_ignored_ddl/run.sh (hunk 1 above)
Comment (lance6716) [reply dm/tests/tracker_ignored_ddl/run.sh:50]: I want to wait some time to see if an error is going to happen

File: dm/syncer/syncer.go (hunk 3 above)
Comment (lance6716) [reply dm/syncer/syncer.go:3315]: can't reproduce

```diff
diff --git a/dm/tests/tracker_ignored_ddl/conf/dm-task.yaml b/dm/tests/tracker_ignored_ddl/conf/dm-task.yaml
index a1dc02ebc..cabf94954 100644
--- a/dm/tests/tracker_ignored_ddl/conf/dm-task.yaml
+++ b/dm/tests/tracker_ignored_ddl/conf/dm-task.yaml
@@ -16,7 +16,7 @@ mysql-instances:
     mydumper-config-name: "global"
     loader-config-name: "global"
     syncer-config-name: "global"
-    filter-rules: ["user-filter-1"]
+    filter-rules: ["user-filter-1", "user-filter-2"]
 
 black-white-list: # compatible with deprecated config
   instance:
@@ -28,6 +28,11 @@ filters:                     # filter rules, mysql instance can ref rules in it
     table-pattern: "*"        # pattern of the upstream table name, wildcard characters (*?) are supported
     events: ["alter table"]  # ignore alter table ddl
     action: Ignore
+  user-filter-2:
+    schema-pattern: "tracker_ignored_ddl"
+    table-pattern: "down"
+    events: ["all ddl"]
+    action: Ignore
 
 mydumpers:
   global:
```

```diff
diff --git a/dm/tests/tracker_ignored_ddl/run.sh b/dm/tests/tracker_ignored_ddl/run.sh
index 3fd1d611e..1882dfec5 100644
--- a/dm/tests/tracker_ignored_ddl/run.sh
+++ b/dm/tests/tracker_ignored_ddl/run.sh
@@ -62,6 +62,42 @@ function run() {
        run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
                "query-status test" \
                "\"stage\": \"Running\"" 1
+
+       run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+               "stop-task test" \
+               "\"result\": true" 2
+
+       # to make use of checkpoint like an incremental task
+       test_track_downstream
+}
+
+function test_track_downstream() {
+  echo ">>>>>>>>>>>>>enter test_track_downstream"
+       run_sql_source1 "create table tracker_ignored_ddl.down (c int primary key, c2 int);"
+       run_sql_source1 "insert into tracker_ignored_ddl.down values (1, 1);"
+
+       # downstream created wrong table
+       run_sql_tidb "create table tracker_ignored_ddl.down (c int primary key);"
+
+  run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "start-task $cur/conf/dm-task.yaml"
+
+       run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+               "query-status test" \
+               "Error 1054: Unknown column" 1
+
+       run_sql_tidb "alter table tracker_ignored_ddl.down add column c2 int;"
+       run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+               "resume-task test" \
+               "\"result\": true" 2
+       sleep 3
+       run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+               "query-status test" \
+               "\"stage\": \"Running\"" 1
+
+  read -p 12312123123
+       cleanup_data $TEST_NAME
+       cleanup_process
 }
 
 cleanup_data $TEST_NAME
```

Comment (D3Hunter) [reply dm/syncer/syncer.go:3317]: https://github.com/pingcap/tiflow/blob/a74c6110c658f199fba88738dc215ec741ef4d91/dm/syncer/checkpoint.go#L935 will cleanup schema-tracker table without checkpoint, so no such issue.

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**